### PR TITLE
Move client only primitives to another dir

### LINKED ISSUE #9218: Establish which crates are client only in the primatives dir
Establish which crates are client only in the primatives dir so that we can separate them out so that it's clearer as which the frame/pallet used primatives are.

Update:

The following primitive crates seem to have no direct calls from frame/*:

- allocator
- block-builder
- chain-spec
- offchain

The following didn't have direct frame dependencies but were referenced by primitives so they are staying put:
- database
- debug-derive
- maybe-compressed-blob
- panic-handler
- serializer
- test-primatives
- utils
- wasm-interface

### PR DESCRIPTION
**Pull request by gilescope**
Fixes #9218 Primitives only used by client moving to under `client/primitives`

### CODE REVIEW

File: bin/node/runtime/Cargo.toml [hunk 1]
```diff
@@ -22,10 +22,10 @@ log = { version = "0.4.14", default-features = false }
 # primitives
 sp-authority-discovery = { version = "3.0.0", default-features = false, path = "../../../primitives/authority-discovery" }
 sp-consensus-babe = { version = "0.9.0", default-features = false, path = "../../../primitives/consensus/babe" }
-sp-block-builder = { path = "../../../primitives/block-builder", default-features = false, version = "3.0.0"}
+sp-block-builder = { path = "../../../client/primitives/block-builder", default-features = false, version = "3.0.0"}
```
[the remark this answers is not in the document]
Comment (bkchr) [reply bin/node/runtime/Cargo.toml:25]: Exactly

File: bin/node/runtime/Cargo.toml [hunk 2]
```diff
@@ -22,10 +22,10 @@ log = { version = "0.4.14", default-features = false }
...
 sp-inherents = { version = "3.0.0", default-features = false, path = "../../../primitives/inherents" }
 node-primitives = { version = "2.0.0", default-features = false, path = "../primitives" }
-sp-offchain = { version = "3.0.0", default-features = false, path = "../../../primitives/offchain" }
+sp-offchain = { version = "3.0.0", default-features = false, path = "../../../client/primitives/offchain" }
```
Comment (bkchr) [inline bin/node/runtime/Cargo.toml:28]: Same

**Review by bkchr — CHANGES_REQUESTED** [review]
`sp-block-builder` and `sp-offchain` are referenced by runtimes, this means they should stay in primitives.

`sp-allocator` could be renamed to `sc-allocator` and moved.
`sp-chain-spec` can just be merged with `sc-chain-spec` then.

**Comment by expenses** [comment]
We probably want a `readme.md` in `client/primitives` that simply says something like hey, there are the client-only primitives.

**Review by expenses — COMMENTED** [review]
I'm happy to go with this when the other reviewers are satisfied

**Comment by gilescope** [comment]
(Might have squashed a couple of warnings of unneeded `&mut` while I was passing as they were annoying me)

File: bin/node-template/node/Cargo.toml [hunk 3]
```diff
@@ -43,7 +43,7 @@ sc-rpc = { version = "3.0.0", path = "../../../client/rpc" }
 sp-api = { version = "3.0.0", path = "../../../primitives/api" }
 sc-rpc-api = { version = "0.9.0", path = "../../../client/rpc-api" }
 sp-blockchain = { version = "3.0.0", path = "../../../primitives/blockchain" }
-sp-block-builder = { version = "3.0.0", path = "../../../primitives/block-builder" }
+sp-block-builder = { version = "3.0.0", path = "../../../client/primitives/block-builder" }
```
[the remark this answers is not in the document]
Comment (gilescope) [reply bin/node-template/node/Cargo.toml]: Have renamed to sc-allocator-primitives and sc-chain-spec-primatives.

File: bin/node/runtime/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gilescope) [reply bin/node/runtime/Cargo.toml]: Sorry - didn't realise. Fixed.

File: client/executor/runtime-test/src/lib.rs [hunk 4]
```diff
@@ -286,9 +286,9 @@ sp_core::wasm_export_functions! {
 		run().is_some()
 	}
 
-	// Just some test to make sure that `sp-allocator` compiles on `no_std`.
-	fn test_sp_allocator_compiles() {
-		sp_allocator::FreeingBumpHeapAllocator::new(0);
+	// Just some test to make sure that `sc-allocator-primitives` compiles on `no_std`.
+	fn test_sc_allocator_primitive_compiles() {
```
Comment (bkchr) [inline client/executor/runtime-test/src/lib.rs:290]: This makes no sense anymore. Please remove it.

File: client/primitives/chain-spec/Cargo.toml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "sp-chain-spec"
+name = "sc-chain-spec-primitives"
```
Comment (bkchr) [inline client/primitives/chain-spec/Cargo.toml:2]: Please move all the code just to `sc-chain-spec`.

**Review by tomusdrw — APPROVED** [review]

File: client/executor/runtime-test/src/lib.rs (hunk 4 above)
Comment (bkchr) [reply client/executor/runtime-test/src/lib.rs:290]: The crate itself doesn't need to be `no_std` anymore.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: client/executor/runtime-test/src/lib.rs (hunk 4 above)
Comment (gilescope) [reply client/executor/runtime-test/src/lib.rs:290]: Why did it need to be no_std in the first place? (Just asking in case I'm missing something)

Comment (bkchr) [reply client/executor/runtime-test/src/lib.rs:290]: There was once the plan to use it from wasm, but this was thrown away. So, it is not needed anymore.

Comment (gilescope) [reply client/executor/runtime-test/src/lib.rs]: ok crate is standardised.

**Comment by gilescope** [comment]
@bkchr any further issues or are we all good now?

**Review by bkchr — APPROVED** [review]

**Comment by trevor-crypto** [comment]
This introduces a breaking change for me because it introduces `async-std` somehow and I am on a non-async platform. The commit before this one works fine. Anyone know why this happened and if it is by design?

**Comment by gilescope** [comment]
@trevor-crypto sorry I've caused you pain! I believe the substrate socket handling at the moment is done via asnyc-std with a tokio executor. (Rust's async story is as ever evolving)

When you say you're on a `non-async` platform can you elaborate on that? (Also always keen to hear a little context of what people are using substrate for - if you can say).

**Comment by trevor-crypto** [comment]
@gilescope No problem. I was only using one type from `sc-rpc-api` -- `ReadProof<Hash>`. I simply copied it and made a note, so all is working again.

I am compiling for intel-sgx, but the target/toolchain does not currently support async/await (fortanix). Not doing anything special, just making modifications to `substrate-api-client`, the unofficial client lib which is also non-async, to use within a product

**Pull request merged**